File: netflix-movie-library-service/api/services/redis_operations_service.py
"""
Redis Operations Service for Netflix Movie Library Explorer
Handles system monitoring, health checks, and operations data using Redis DB 2.
"""

import redis
import json
import time
import psutil
from typing import Dict, Any, List, Optional
from loguru import logger
from .redis_service import redis_service

class RedisOperationsService:
    """Service for managing operations data in Redis DB 2."""
    
    def __init__(self):
        """Initialize the operations service."""
        self.redis = redis_service.get_operations_db()
        self.metrics = {
            "system": {},
            "performance": {},
            "alerts": []
        }
        logger.info("Redis Operations Service initialized")
    
    def collect_system_metrics(self) -> bool:
        """Collect current system metrics."""
        try:
            timestamp = int(time.time() * 1000)  # Milliseconds
            
            # Collect CPU usage
            cpu_usage = psutil.cpu_percent(interval=1)
            
            # Collect memory usage
            memory_info = psutil.virtual_memory()
            memory_usage = memory_info.percent
            
            # Collect Redis memory usage
            redis_memory = self.get_redis_memory_usage()
            
            # Store in time series (if RedisTimeSeries is available)
            try:
                self.redis.execute_command("TS.ADD", "system_metrics:cpu_usage", timestamp, cpu_usage,
                                         "LABELS", "type", "cpu")
                self.redis.execute_command("TS.ADD", "system_metrics:memory_usage", timestamp, memory_usage,
                                         "LABELS", "type", "memory")
                self.redis.execute_command("TS.ADD", "system_metrics:redis_memory", timestamp, redis_memory,
                                         "LABELS", "type", "redis")
            except:
                # Fallback to regular Redis operations
                self.redis.lpush(f"system_metrics:cpu_usage:{int(time.time())}", 
                               json.dumps({"timestamp": timestamp, "value": cpu_usage}))
                self.redis.lpush(f"system_metrics:memory_usage:{int(time.time())}", 
                               json.dumps({"timestamp": timestamp, "value": memory_usage}))
                self.redis.lpush(f"system_metrics:redis_memory:{int(time.time())}", 
                               json.dumps({"timestamp": timestamp, "value": redis_memory}))
            
            # Update current metrics
            self.metrics["system"] = {
                "cpu": cpu_usage,
                "memory": memory_usage,
                "redis": redis_memory,
                "timestamp": time.time()
            }
            
            logger.debug(f"📊 Collected system metrics: CPU={cpu_usage}%, Memory={memory_usage}%, Redis={redis_memory}%")
            return True
            
        except Exception as e:
            logger.error(f"Error collecting system metrics: {e}")
            return False
    
    def collect_performance_metrics(self, response_time: float, error_rate: float) -> bool:
        """Collect application performance metrics."""
        try:
            timestamp = int(time.time() * 1000)  # Milliseconds
            
            # Store in time series (if RedisTimeSeries is available)
            try:
                self.redis.execute_command("TS.ADD", "app_metrics:response_time", timestamp, response_time,
                                         "LABELS", "type", "performance")
                self.redis.execute_command("TS.ADD", "app_metrics:error_rate", timestamp, error_rate,
                                         "LABELS", "type", "errors")
            except:
                # Fallback to regular Redis operations
                self.redis.lpush(f"app_metrics:response_time:{int(time.time())}", 
                               json.dumps({"timestamp": timestamp, "value": response_time}))
                self.redis.lpush(f"app_metrics:error_rate:{int(time.time())}", 
                               json.dumps({"timestamp": timestamp, "value": error_rate}))
            
            # Update current metrics
            self.metrics["performance"] = {
                "responseTime": response_time,
                "errorRate": error_rate,
                "timestamp": time.time()
            }
            
            logger.debug(f"📊 Collected performance metrics: ResponseTime={response_time}ms, ErrorRate={error_rate}%")
            return True
            
        except Exception as e:
            logger.error(f"Error collecting performance metrics: {e}")
            return False
    
    def check_alerts(self) -> List[Dict[str, Any]]:
        """Check for system alerts based on thresholds."""
        try:
            alerts = []
            current_time = time.time()
            
            # Get configuration
            config = self.get_configuration()
            thresholds = config["systemThresholds"]
            perf_thresholds = config["performanceThresholds"]
            
            # Check CPU threshold
            if self.metrics["system"].get("cpu", 0) > thresholds["cpu"]:
                alerts.append({
                    "type": "warning",
                    "message": f"High CPU usage: {self.metrics['system'].get('cpu', 0):.1f}%",
                    "timestamp": current_time,
                    "severity": "high",
                    "metric": "cpu",
                    "value": self.metrics["system"].get("cpu", 0),
                    "threshold": thresholds["cpu"]
                })
            
            # Check memory threshold
            if self.metrics["system"].get("memory", 0) > thresholds["memory"]:
                alerts.append({
                    "type": "warning",
                    "message": f"High memory usage: {self.metrics['system'].get('memory', 0):.1f}%",
                    "timestamp": current_time,
                    "severity": "high",
                    "metric": "memory",
                    "value": self.metrics["system"].get("memory", 0),
                    "threshold": thresholds["memory"]
                })
            
            # Check Redis memory threshold
            if self.metrics["system"].get("redis", 0) > thresholds["redis"]:
                alerts.append({
                    "type": "warning",
                    "message": f"High Redis memory usage: {self.metrics['system'].get('redis', 0):.1f}%",
                    "timestamp": current_time,
                    "severity": "critical",
                    "metric": "redis",
                    "value": self.metrics["system"].get("redis", 0),
                    "threshold": thresholds["redis"]
                })
            
            # Check response time threshold
            if self.metrics["performance"].get("responseTime", 0) > perf_thresholds["responseTime"]:
                alerts.append({
                    "type": "warning",
                    "message": f"High response time: {self.metrics['performance'].get('responseTime', 0):.1f}ms",
                    "timestamp": current_time,
                    "severity": "medium",
                    "metric": "response_time",
                    "value": self.metrics["performance"].get("responseTime", 0),
                    "threshold": perf_thresholds["responseTime"]
                })
            
            # Check error rate threshold
            if self.metrics["performance"].get("errorRate", 0) > perf_thresholds["errorRate"]:
                alerts.append({
                    "type": "error",
                    "message": f"High error rate: {self.metrics['performance'].get('errorRate', 0):.1f}%",
                    "timestamp": current_time,
                    "severity": "critical",
                    "metric": "error_rate",
                    "value": self.metrics["performance"].get("errorRate", 0),
                    "threshold": perf_thresholds["errorRate"]
                })
            
            # Store alerts
            if alerts:
                self.metrics["alerts"] = alerts
                for alert in alerts:
                    self.redis.lpush("operations:alerts", json.dumps(alert))
                
                # Keep only last 50 alerts
                self.redis.ltrim("operations:alerts", 0, 49)
            
            return alerts
            
        except Exception as e:
            logger.error(f"Error checking alerts: {e}")
            return []
    
    def get_system_metrics(self) -> Dict[str, Any]:
        """Get system metrics for Operations dashboard."""
        try:
            # Get recent system metrics (last hour)
            one_hour_ago = int((time.time() - 3600) * 1000)
            current_time = int(time.time() * 1000)
            
            # Try to get time series data
            try:
                cpu_data = self.redis.execute_command("TS.RANGE", "system_metrics:cpu_usage", one_hour_ago, current_time)
                memory_data = self.redis.execute_command("TS.RANGE", "system_metrics:memory_usage", one_hour_ago, current_time)
                redis_data = self.redis.execute_command("TS.RANGE", "system_metrics:redis_memory", one_hour_ago, current_time)
                
                return {
                    "cpu": {
                        "current": self.metrics["system"].get("cpu", 0),
                        "history": [{"timestamp": point[0], "value": point[1]} for point in cpu_data]
                    },
                    "memory": {
                        "current": self.metrics["system"].get("memory", 0),
                        "history": [{"timestamp": point[0], "value": point[1]} for point in memory_data]
                    },
                    "redis": {
                        "current": self.metrics["system"].get("redis", 0),
                        "history": [{"timestamp": point[0], "value": point[1]} for point in redis_data]
                    },
                    "timestamp": self.metrics["system"].get("timestamp", time.time())
                }
            except:
                # Fallback to regular Redis operations
                return {
                    "cpu": {
                        "current": self.metrics["system"].get("cpu", 0),
                        "history": []
                    },
                    "memory": {
                        "current": self.metrics["system"].get("memory", 0),
                        "history": []
                    },
                    "redis": {
                        "current": self.metrics["system"].get("redis", 0),
                        "history": []
                    },
                    "timestamp": self.metrics["system"].get("timestamp", time.time())
                }
                
        except Exception as e:
            logger.error(f"Error getting system metrics: {e}")
            return {
                "cpu": {"current": 0, "history": []},
                "memory": {"current": 0, "history": []},
                "redis": {"current": 0, "history": []},
                "timestamp": time.time()
            }
    
    def get_performance_metrics(self) -> Dict[str, Any]:
        """Get performance metrics for Operations dashboard."""
        try:
            # Get recent performance metrics (last hour)
            one_hour_ago = int((time.time() - 3600) * 1000)
            current_time = int(time.time() * 1000)
            
            # Try to get time series data
            try:
                response_time_data = self.redis.execute_command("TS.RANGE", "app_metrics:response_time", one_hour_ago, current_time)
                error_rate_data = self.redis.execute_command("TS.RANGE", "app_metrics:error_rate", one_hour_ago, current_time)
                
                return {
                    "responseTime": {
                        "current": self.metrics["performance"].get("responseTime", 0),
                        "history": [{"timestamp": point[0], "value": point[1]} for point in response_time_data]
                    },
                    "errorRate": {
                        "current": self.metrics["performance"].get("errorRate", 0),
                        "history": [{"timestamp": point[0], "value": point[1]} for point in error_rate_data]
                    },
                    "timestamp": self.metrics["performance"].get("timestamp", time.time())
                }
            except:
                # Fallback to regular Redis operations
                return {
                    "responseTime": {
                        "current": self.metrics["performance"].get("responseTime", 0),
                        "history": []
                    },
                    "errorRate": {
                        "current": self.metrics["performance"].get("errorRate", 0),
                        "history": []
                    },
                    "timestamp": self.metrics["performance"].get("timestamp", time.time())
                }
                
        except Exception as e:
            logger.error(f"Error getting performance metrics: {e}")
            return {
                "responseTime": {"current": 0, "history": []},
                "errorRate": {"current": 0, "history": []},
                "timestamp": time.time()
            }
    
    def get_alerts(self) -> List[Dict[str, Any]]:
        """Get recent alerts."""
        try:
            alerts = self.redis.lrange("operations:alerts", 0, 49)
            return [json.loads(alert) for alert in alerts]
        except Exception as e:
            logger.error(f"Error getting alerts: {e}")
            return []
    
    def get_redis_info(self) -> Dict[str, Any]:
        """Get Redis information."""
        try:
            info = self.redis.info()
            
            # Parse Redis INFO command output
            parsed_info = {}
            for line in info.split('\n'):
                if ':' in line:
                    key, value = line.split(':', 1)
                    parsed_info[key] = value.strip()
            
            return {
                "version": parsed_info.get("redis_version", "Unknown"),
                "uptime": int(parsed_info.get("uptime_in_seconds", 0)),
                "connectedClients": int(parsed_info.get("connected_clients", 0)),
                "usedMemory": parsed_info.get("used_memory_human", "0B"),
                "totalCommands": int(parsed_info.get("total_commands_processed", 0)),
                "keyspaceHits": int(parsed_info.get("keyspace_hits", 0)),
                "keyspaceMisses": int(parsed_info.get("keyspace_misses", 0)),
                "timestamp": time.time()
            }
        except Exception as e:
            logger.error(f"Error getting Redis info: {e}")
            return {
                "version": "Unknown",
                "uptime": 0,
                "connectedClients": 0,
                "usedMemory": "0B",
                "totalCommands": 0,
                "keyspaceHits": 0,
                "keyspaceMisses": 0,
                "timestamp": time.time()
            }
    
    def get_configuration(self) -> Dict[str, Any]:
        """Get operations configuration."""
        try:
            config = self.redis.hgetall("operations:configuration")
            
            return {
                "systemThresholds": {
                    "cpu": int(config.get("cpu_threshold", 80)),
                    "memory": int(config.get("memory_threshold", 85)),
                    "redis": int(config.get("redis_threshold", 90))
                },
                "performanceThresholds": {
                    "responseTime": int(config.get("response_time_threshold", 2000)),
                    "errorRate": int(config.get("error_rate_threshold", 5))
                },
                "alerting": {
                    "enabled": config.get("alerting_enabled", "true") == "true",
                    "email": config.get("alert_email", ""),
                    "webhook": config.get("alert_webhook", "")
                },
                "retention": {
                    "metrics": int(config.get("metrics_retention_days", 7)),
                    "logs": int(config.get("logs_retention_days", 30)),
                    "alerts": int(config.get("alerts_retention_days", 90))
                }
            }
        except Exception as e:
            logger.error(f"Error getting configuration: {e}")
            return {
                "systemThresholds": {"cpu": 80, "memory": 85, "redis": 90},
                "performanceThresholds": {"responseTime": 2000, "errorRate": 5},
                "alerting": {"enabled": True, "email": "", "webhook": ""},
                "retention": {"metrics": 7, "logs": 30, "alerts": 90}
            }
    
    def update_configuration(self, config: Dict[str, Any]) -> bool:
        """Update operations configuration."""
        try:
            self.redis.hset("operations:configuration", mapping={
                "cpu_threshold": str(config["systemThresholds"]["cpu"]),
                "memory_threshold": str(config["systemThresholds"]["memory"]),
                "redis_threshold": str(config["systemThresholds"]["redis"]),
                "response_time_threshold": str(config["performanceThresholds"]["responseTime"]),
                "error_rate_threshold": str(config["performanceThresholds"]["errorRate"]),
                "alerting_enabled": str(config["alerting"]["enabled"]),
                "alert_email": config["alerting"]["email"],
                "alert_webhook": config["alerting"]["webhook"],
                "metrics_retention_days": str(config["retention"]["metrics"]),
                "logs_retention_days": str(config["retention"]["logs"]),
                "alerts_retention_days": str(config["retention"]["alerts"])
            })
            
            logger.info("Configuration updated successfully")
            return True
            
        except Exception as e:
            logger.error(f"Error updating configuration: {e}")
            return False
    
    def get_redis_memory_usage(self) -> float:
        """Get Redis memory usage percentage."""
        try:
            info = self.redis.info()
            used_memory = int(info.get("used_memory", 0))
            max_memory = int(info.get("maxmemory", 0))
            
            if max_memory > 0:
                return (used_memory / max_memory) * 100
            else:
                return 0.0
        except Exception as e:
            logger.error(f"Error getting Redis memory usage: {e}")
            return 0.0
    
    def health_check(self) -> Dict[str, Any]:
        """Check operations service health."""
        try:
            self.redis.ping()
            return {
                "status": "healthy",
                "message": "Redis Operations Service is operational",
                "metrics": self.metrics,
                "timestamp": time.time()
            }
        except Exception as e:
            return {
                "status": "error",
                "message": f"Redis Operations Service error: {str(e)}",
                "timestamp": time.time()
            }

# Create singleton instance
redis_operations_service = RedisOperationsService()

File: netflix-movie-library-ui/src/services/redisOperationsService.js
/**
 * Redis Operations Service
 * Handles system monitoring, health checks, and operations data for Operations dashboard
 */

class RedisOperationsService {
  constructor() {
    this.redis = null
    this.isConnected = false
    this.metrics = {
      system: {},
      performance: {},
      alerts: []
    }
    this.init()
  }

  async init() {
    try {
      this.redis = await this.connectRedis()
      this.isConnected = true
      console.log('✅ Redis Operations Service initialized')
      
      // Start metrics collection
      this.startMetricsCollection()
    } catch (error) {
      console.error('❌ Failed to initialize Redis Operations Service:', error)
      this.isConnected = false
    }
  }

  async connectRedis() {
    // Connect to the backend API for Redis operations
    const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
    
    return {
      // Health check
      ping: async () => {
        try {
          const response = await fetch(`${API_BASE_URL}/health`)
          const data = await response.json()
          return data.status === 'healthy' ? 'PONG' : 'ERROR'
        } catch (error) {
          console.error('Error pinging Redis:', error)
          return 'ERROR'
        }
      },
      
      // Get Redis info
      info: async () => {
        try {
          const response = await fetch(`${API_BASE_URL}/api/operations/redis-info`)
          const data = await response.json()
          if (data.success) {
            return `redis_version:${data.data.version}\nused_memory:${data.data.usedMemory}\nconnected_clients:${data.data.connectedClients}`
          }
          return 'redis_version:Unknown\nused_memory:0B\nconnected_clients:0'
        } catch (error) {
          console.error('Error getting Redis info:', error)
          return 'redis_version:Unknown\nused_memory:0B\nconnected_clients:0'
        }
      },
      
      // Get configuration
      hgetall: async (key) => {
        try {
          const response = await fetch(`${API_BASE_URL}/api/operations/configuration`)
          const data = await response.json()
          return data.success ? data.data : {}
        } catch (error) {
          console.error('Error getting configuration:', error)
          return {}
        }
      },
      
      // Update configuration
      hset: async (key, field, value) => {
        try {
          const response = await fetch(`${API_BASE_URL}/api/operations/configuration`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ [field]: value })
          })
          return response.ok ? 1 : 0
        } catch (error) {
          console.error('Error updating configuration:', error)
          return 0
        }
      },
      
      // Time series operations (fallback to regular operations)
      ts_add: async (key, timestamp, value, labels) => {
        console.log(`Time series tracking: ${key} ${timestamp} ${value}`, labels)
        return Promise.resolve(timestamp)
      },
      
      ts_range: async (key, from, to) => {
        console.log(`Time series query: ${key} ${from} ${to}`)
        return Promise.resolve([])
      },
      
      // List operations
      lpush: async (key, value) => {
        console.log(`List push: ${key} ${value}`)
        return Promise.resolve(1)
      },
      
      lrange: async (key, start, stop) => {
        console.log(`List range: ${key} ${start} ${stop}`)
        return Promise.resolve([])
      },
      
      // Sorted set operations
      zadd: async (key, score, member) => {
        console.log(`Sorted set add: ${key} ${score} ${member}`)
        return Promise.resolve(1)
      },
      
      zrevrange: async (key, start, stop) => {
        console.log(`Sorted set range: ${key} ${start} ${stop}`)
        return Promise.resolve([])
      }
    }
  }

  // ==================== SYSTEM METRICS COLLECTION ====================

  startMetricsCollection() {
    // Collect system metrics every 30 seconds
    setInterval(() => {
      this.collectSystemMetrics()
    }, 30000)

    // Collect performance metrics every 10 seconds
    setInterval(() => {
      this.collectPerformanceMetrics()
    }, 10000)

    // Check alerts every 60 seconds
    setInterval(() => {
      this.checkAlerts()
    }, 60000)
  }

  async collectSystemMetrics() {
    if (!this.isConnected) return

    try {
      const timestamp = Date.now()
      
      // Collect CPU usage (mock)
      const cpuUsage = this.getMockCpuUsage()
      await this.redis.ts_add('system_metrics:cpu_usage', timestamp, cpuUsage, {
        type: 'cpu'
      })

      // Collect memory usage (mock)
      const memoryUsage = this.getMockMemoryUsage()
      await this.redis.ts_add('system_metrics:memory_usage', timestamp, memoryUsage, {
        type: 'memory'
      })

      // Collect Redis memory usage
      const redisMemory = await this.getRedisMemoryUsage()
      await this.redis.ts_add('system_metrics:redis_memory', timestamp, redisMemory, {
        type: 'redis'
      })

      // Update current metrics
      this.metrics.system = {
        cpu: cpuUsage,
        memory: memoryUsage,
        redis: redisMemory,
        timestamp: new Date().toISOString()
      }

    } catch (error) {
      console.error('Error collecting system metrics:', error)
    }
  }

  async collectPerformanceMetrics() {
    if (!this.isConnected) return

    try {
      const timestamp = Date.now()
      
      // Collect response time (mock)
      const responseTime = this.getMockResponseTime()
      await this.redis.ts_add('app_metrics:response_time', timestamp, responseTime, {
        type: 'performance'
      })

      // Collect error rate (mock)
      const errorRate = this.getMockErrorRate()
      await this.redis.ts_add('app_metrics:error_rate', timestamp, errorRate, {
        type: 'errors'
      })

      // Update current metrics
      this.metrics.performance = {
        responseTime: responseTime,
        errorRate: errorRate,
        timestamp: new Date().toISOString()
      }

    } catch (error) {
      console.error('Error collecting performance metrics:', error)
    }
  }

  // ==================== ALERTING SYSTEM ====================

  async checkAlerts() {
    if (!this.isConnected) return

    try {
      const alerts = []

      // Check CPU threshold
      if (this.metrics.system.cpu > 80) {
        alerts.push({
          type: 'warning',
          message: `High CPU usage: ${this.metrics.system.cpu}%`,
          timestamp: new Date().toISOString(),
          severity: 'high'
        })
      }

      // Check memory threshold
      if (this.metrics.system.memory > 85) {
        alerts.push({
          type: 'warning',
          message: `High memory usage: ${this.metrics.system.memory}%`,
          timestamp: new Date().toISOString(),
          severity: 'high'
        })
      }

      // Check Redis memory threshold
      if (this.metrics.system.redis > 90) {
        alerts.push({
          type: 'warning',
          message: `High Redis memory usage: ${this.metrics.system.redis}%`,
          timestamp: new Date().toISOString(),
          severity: 'critical'
        })
      }

      // Check response time threshold
      if (this.metrics.performance.responseTime > 2000) {
        alerts.push({
          type: 'warning',
          message: `High response time: ${this.metrics.performance.responseTime}ms`,
          timestamp: new Date().toISOString(),
          severity: 'medium'
        })
      }

      // Check error rate threshold
      if (this.metrics.performance.errorRate > 5) {
        alerts.push({
          type: 'error',
          message: `High error rate: ${this.metrics.performance.errorRate}%`,
          timestamp: new Date().toISOString(),
          severity: 'critical'
        })
      }

      // Store alerts
      if (alerts.length > 0) {
        this.metrics.alerts = [...this.metrics.alerts, ...alerts].slice(-50) // Keep last 50 alerts
        
        // Store in Redis
        for (const alert of alerts) {
          await this.redis.lpush('operations:alerts', JSON.stringify(alert))
        }
      }

    } catch (error) {
      console.error('Error checking alerts:', error)
    }
  }

  // ==================== OPERATIONS DATA RETRIEVAL ====================

  async getSystemMetrics() {
    if (!this.isConnected) return this.getMockSystemMetrics()

    try {
      const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
      const response = await fetch(`${API_BASE_URL}/api/operations/system-metrics`)
      const data = await response.json()
      
      if (data.success) {
        return data.data
      } else {
        console.error('Error getting system metrics:', data.error)
        return this.getMockSystemMetrics()
      }
    } catch (error) {
      console.error('Error getting system metrics:', error)
      return this.getMockSystemMetrics()
    }
  }

  async getPerformanceMetrics() {
    if (!this.isConnected) return this.getMockPerformanceMetrics()

    try {
      const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
      const response = await fetch(`${API_BASE_URL}/api/operations/performance-metrics`)
      const data = await response.json()
      
      if (data.success) {
        return data.data
      } else {
        console.error('Error getting performance metrics:', data.error)
        return this.getMockPerformanceMetrics()
      }
    } catch (error) {
      console.error('Error getting performance metrics:', error)
      return this.getMockPerformanceMetrics()
    }
  }

  async getAlerts() {
    if (!this.isConnected) return this.getMockAlerts()

    try {
      const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
      const response = await fetch(`${API_BASE_URL}/api/operations/alerts`)
      const data = await response.json()
      
      if (data.success) {
        return data.data
      } else {
        console.error('Error getting alerts:', data.error)
        return this.getMockAlerts()
      }
    } catch (error) {
      console.error('Error getting alerts:', error)
      return this.getMockAlerts()
    }
  }

  async getRedisInfo() {
    if (!this.isConnected) return this.getMockRedisInfo()

    try {
      const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
      const response = await fetch(`${API_BASE_URL}/api/operations/redis-info`)
      const data = await response.json()
      
      if (data.success) {
        return data.data
      } else {
        console.error('Error getting Redis info:', data.error)
        return this.getMockRedisInfo()
      }
    } catch (error) {
      console.error('Error getting Redis info:', error)
      return this.getMockRedisInfo()
    }
  }

  // ==================== CONFIGURATION MANAGEMENT ====================

  async getConfiguration() {
    if (!this.isConnected) return this.getMockConfiguration()

    try {
      const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
      const response = await fetch(`${API_BASE_URL}/api/operations/configuration`)
      const data = await response.json()
      
      if (data.success) {
        return data.data
      } else {
        console.error('Error getting configuration:', data.error)
        return this.getMockConfiguration()
      }
    } catch (error) {
      console.error('Error getting configuration:', error)
      return this.getMockConfiguration()
    }
  }

  async updateConfiguration(config) {
    if (!this.isConnected) return false

    try {
      const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
      const response = await fetch(`${API_BASE_URL}/api/operations/configuration`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(config)
      })
      const data = await response.json()
      
      return data.success
    } catch (error) {
      console.error('Error updating configuration:', error)
      return false
    }
  }

  // ==================== UTILITY METHODS ====================

  async getRedisMemoryUsage() {
    try {
      const info = await this.redis.info()
      const memoryMatch = info.match(/used_memory:(\d+)/)
      const maxMemoryMatch = info.match(/maxmemory:(\d+)/)
      
      if (memoryMatch && maxMemoryMatch) {
        const used = parseInt(memoryMatch[1])
        const max = parseInt(maxMemoryMatch[1])
        return max > 0 ? Math.round((used / max) * 100) : 0
      }
      
      return 0
    } catch (error) {
      return 0
    }
  }

  // ==================== MOCK DATA FOR DEVELOPMENT ====================

  getMockCpuUsage() {
    return Math.floor(Math.random() * 30) + 20 // 20-50%
  }

  getMockMemoryUsage() {
    return Math.floor(Math.random() * 40) + 30 // 30-70%
  }

  getMockResponseTime() {
    return Math.floor(Math.random() * 500) + 100 // 100-600ms
  }

  getMockErrorRate() {
    return Math.floor(Math.random() * 3) // 0-3%
  }

  getMockSystemMetrics() {
    return {
      cpu: { current: 35, history: [] },
      memory: { current: 45, history: [] },
      redis: { current: 25, history: [] },
      timestamp: new Date().toISOString()
    }
  }

  getMockPerformanceMetrics() {
    return {
      responseTime: { current: 250, history: [] },
      errorRate: { current: 1, history: [] },
      timestamp: new Date().toISOString()
    }
  }

  getMockAlerts() {
    return [
      {
        type: 'info',
        message: 'System running normally',
        timestamp: new Date().toISOString(),
        severity: 'low'
      }
    ]
  }

  getMockRedisInfo() {
    return {
      version: '7.0.0',
      uptime: 86400,
      connectedClients: 5,
      usedMemory: '1.2MB',
      totalCommands: 15000,
      keyspaceHits: 12000,
      keyspaceMisses: 3000,
      timestamp: new Date().toISOString()
    }
  }

  getMockConfiguration() {
    return {
      systemThresholds: {
        cpu: 80,
        memory: 85,
        redis: 90
      },
      performanceThresholds: {
        responseTime: 2000,
        errorRate: 5
      },
      alerting: {
        enabled: true,
        email: 'admin@example.com',
        webhook: ''
      },
      retention: {
        metrics: 7,
        logs: 30,
        alerts: 90
      }
    }
  }

  // ==================== HEALTH CHECK ====================

  async healthCheck() {
    try {
      if (!this.isConnected) {
        return { status: 'disconnected', message: 'Redis not connected' }
      }

      await this.redis.ping()
      
      return { 
        status: 'healthy', 
        message: 'Redis Operations Service is operational',
        metrics: this.metrics,
        timestamp: new Date().toISOString()
      }
    } catch (error) {
      return { 
        status: 'error', 
        message: `Redis Operations Service error: ${error.message}`,
        timestamp: new Date().toISOString()
      }
    }
  }
}

// Export singleton instance
export default new RedisOperationsService()
